### PR TITLE
[BugFix] Fix incompatible zonemap reuse for fast schema evolution in shared-data

### PR DESCRIPTION
**Pull request by banmoy**
## Why I'm doing:
Fast schema evolution in shared-data modifies a column's type without rewriting data files, and the existing zonemap index is reused for the new type. However, this reuse is not always safe. For certain type conversions, the min/max values stored in the zonemap are no longer valid under the new type's comparison rules, leading to incorrect query results.

For example, when changing a column from `INT` to `STRING`, the numeric ordering (`9 < 123`) differs from the lexicographical ordering (`"123" < "9"`). Reusing the integer-based zonemap for string-based comparisons will cause the query engine to prune data incorrectly. This issue can be reproduced with the following steps:
```
create table t (c0 int, c1 int) duplicate key (c0);

insert into t values (1, 123), (2, 9);

select * from t where c1 = 123;             -- expect one row. correct
+------+------+
| c0   | c1   |
+------+------+
|    1 |  123 |
+------+------+
1 row in set (0.02 sec)

alter table t modify column c1 string;

select * from t where c1 = "123";           -- expect one row, but is empty. wrong
Empty set (0.01 sec)
```

From the profile, we can observe that the zonemap filter unexpectedly pruned all rows.

image

This PR wants to fix the issue.

### Conditions for Zonemap Index Reuse
The zonemap index mechanism is briefly illustrated below. The type conversion that wants to safely reuse the existing zonemap index must meet two conditions simultaneously:
* The type conversion must be monotonically non-decreasing. This ensures that the original min/max values remain valid boundaries after the conversion.
* The type conversion must not change non-null values to null. This is required to keep the has_null and has_not_null metadata accurate.

image

Based on these conditions, the following is the compatibility matrix defining which type conversions can safely reuse the zonemap index.
- Schema Change Conversion Prohibited: ⚪
- Zonmap Index can be reused:  ✅
- Zonmap Index can not be reused:  ❌
image

### Solution

To address this issue, two potential solutions were considered:

**Option 1: Prohibit Incompatible Fast Schema Evolution (FE-side Check)**

This approach involves enhancing the validation logic in the Frontend (FE). Before a fast schema evolution is executed, the system checks if the proposed type change is compatible with zonemap index reuse. If the conversion is not on the approved list, fast schema evolution is disallowed, and the operation falls back to the traditional method of rewriting data and rebuilding the index.

   *   **Pros:** Guarantees that zonemap indexes are always valid and can be used by the query optimizer, ensuring consistent and predictable performance.
   *   **Cons:** A subset of type changes (primarily conversions to/from `VARCHAR`) will not be able to use the fast schema evolution feature. Based on experience, these types of conversions are infrequent.

**Option 2: Conditionally Disable Zonemap Index at Query Time (BE-side Logic)**

This alternative would allow all type conversions to proceed with fast schema evolution. The Backend (BE) would then be responsible for detecting incompatible conversions at query time and disabling the use of the zonemap index for those specific columns.

   *   **Pros:** Allows all type modifications to benefit from fast schema evolution, avoiding data rewrites.
   *   **Cons:** May lead to significant and unpredictable performance degradation for queries, as the optimizer would be unable to use the zonemap for pruning.

**Decision**

Query performance is a critical priority. The zonemap index is a widely used and highly effective optimization, and disabling it could have a severe, hard-to-predict impact. Therefore, **choose Option 1** to guarantee index correctness and consistent query performance. The trade-off of disallowing fast schema evolution for a small number of less common type conversions is acceptable to avoid this risk.

### Code Changes
- Add `ZONEMAP_SCHEMA_CHANGE_MATRIX` and `canReuseZonemapIndex(fromType, toType)` to `SchemaChangeTypeCompatibility`, with clear docs on reuse principles.
- Integrate reuse check into FE fast schema evolution path:
  - In schema change planning, for shared-data fast path, validate each altered column via `canReuseZonemapIndex`.
  - If any column is incompatible, disable fast schema evolution and fall back to data rewrite (rebuild zonemap index).
- Tests:
  - Add/extend unit tests for the compatibility matrix and edge cases.
  - Add/adjust schema change tests to verify fallback behavior under incompatible conversions (e.g., VARCHAR↔INT).

Fixes #issue

## What type of PR is this:

- [X] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [X] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [X] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [X] I have checked the version labels which the pr will be auto-backported to the target branch
  - [X] 4.0
  - [X] 3.5
  - [X] 3.4
  - [X] 3.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java [hunk 1]
```diff
@@ -903,6 +903,10 @@ public boolean supportBloomFilter() {
                 !isJsonType() && !isOnlyMetricType() && !isFunctionType() && !isBinaryType();
     }
 
+    public boolean supportZonemap() {
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Type.java]:
```suggestion
    public boolean supportZoneMap() {
```

**Comment by alvin-phoenix-ai** [comment]
@cursor review

File: fe/fe-core/src/main/java/com/starrocks/catalog/Type.java (hunk 1 above)
Comment (banmoy) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Type.java]: fixed

**Review by xiangguangyxg — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Comment by alvin-phoenix-ai** [comment]
@cursor review

**Comment by banmoy** [comment]
ignore merge 3.3.19
